File: memebot/src/notify.ts
import type { Config } from './config';
import { log, type Fetch } from './util';

export type Notifier = (msg: string) => Promise<void>;

/** Escribe en el log y, si está configurado, avisa por Telegram. Nunca lanza errores. */
export function makeNotifier(
  cfg: Pick<Config, 'MODE' | 'TELEGRAM_BOT_TOKEN' | 'TELEGRAM_CHAT_ID'>,
  fetchImpl: Fetch = fetch,
): Notifier {
  return async (msg) => {
    log('info', msg);
    if (!cfg.TELEGRAM_BOT_TOKEN || !cfg.TELEGRAM_CHAT_ID) return;
    try {
      const res = await fetchImpl(`https://api.telegram.org/bot${cfg.TELEGRAM_BOT_TOKEN}/sendMessage`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ chat_id: cfg.TELEGRAM_CHAT_ID, text: `[memebot ${cfg.MODE}] ${msg}` }),
        signal: AbortSignal.timeout(10_000),
      });
      if (!res.ok) log('warn', `Telegram respondió ${res.status}`);
    } catch {
      // No se registra la URL: lleva el token del bot.
      log('warn', 'no se pudo avisar por Telegram');
    }
  };
}

File: memebot/memebot/notify.py
from __future__ import annotations

import httpx

from .config import Config
from .util import log


class Notifier:
    """Escribe en el log y, si está configurado, avisa por Telegram. Nunca lanza errores."""

    def __init__(self, cfg: Config, http: httpx.Client | None = None):
        self.cfg, self.http = cfg, http

    def __call__(self, msg: str) -> None:
        log.info(msg)
        token, chat = self.cfg.TELEGRAM_BOT_TOKEN, self.cfg.TELEGRAM_CHAT_ID
        if not (token and chat and self.http):
            return
        try:
            r = self.http.post(
                f"https://api.telegram.org/bot{token}/sendMessage",
                json={"chat_id": chat, "text": f"[memebot {self.cfg.MODE}] {msg}"},
                timeout=10,
            )
            if r.status_code >= 400:
                log.warning("Telegram respondió %s", r.status_code)
        except Exception:
            # No se registra la URL: lleva el token del bot.
            log.warning("no se pudo avisar por Telegram")
